core/meta/src/TClass.cxx

```diff
@@ -4358,29 +4357,33 @@ Int_t TClass::GetNmethods()
 
 TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
 {
-   TVirtualStreamerInfo *guess = fLastReadInfo;
-   if (guess && guess->GetClassVersion() == version) {
-      // If the StreamerInfo is assigned to the fLastReadInfo, we are
-      // guaranteed it was built and compiled.
-      return guess;
-   }
+   TVirtualStreamerInfo *sinfo = fLastReadInfo;
 
-   R__LOCKGUARD(gInterpreterMutex);
+   // Version 0 is special, it means the currently loaded version.
+   // We need to set it at the beginning to be able to guess it correctly.
 
-   // Handle special version, 0 means currently loaded version.
-   // Warning:  This may be -1 for an emulated class.
-   // If version == -2, the user is requested the emulated streamerInfo
-   // for an abstract base class even though we have a dictionary for it.
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
```

Comment:
> If fClassInfo is not thread-safe
Here we are talking about fClassVersion :)
> At this point, seeing how many locks we have everywhere, I think we should seriously consider using RCU in ROOT instead of locks,
In general as far as I can tell this is indeed what we have been going towards where we can.  When more than one entity need to be updated 'at the same time' (eg. the interpreter states) we have no choice but to take lock ..

>  we need to make it thread-safe or go back to taking the lock, despite the performance penalty. 

Yes my point exactly :)